### PR TITLE
[Frontend] Add support for logical or, and and not in autograph

### PR DESCRIPTION
**Pull request by grwlf**
Support Python and-, or- and not- operators in Autograph-converted programs.

[sc-41303]

Test cases mentioned:

* [x] cases where the boolean statement is a mix of dynamic and static variables,
* [x] test cases where the parameters are of different types (jax arrays, booleans, numpy arrays)
* [x] cases where the boolean is used as conditionals for if statements
* [x] cases where the boolean is used as the conditional for while statemen
* [x] test cases where the parameters are arbitrary Python objects which can decay to Boolean (like strings, lists, etc, anything that Python allows in an if statement)

### CODE REVIEW

**Comment by grwlf** [comment]
Hi @josh146. I expect this PR to have corner-cases, so I 'd like to know what we want to support at the first place here. Do we have any use-case in mind, e.g. among the tutorial algorithms which I could check?

File: frontend/test/pytest/test_autograph.py [hunk 1]
```diff
@@ -1256,6 +1256,87 @@ def f1():
...
+
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+        monkeypatch.setattr("catalyst.ag_primitives._emulate_fallback_errors", True)
+
+        def f1(a):
+            return a > 0.0 and a < 1.0
+
+        with pytest.raises(AutoGraphError, match="Emulated"):
+            qjit(autograph=True)(f1)(0.5)
+
```
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:1338]: Nice tests @grwlf! Is it worth also adding some lit/unit tests, to ensure that `and` and `or` are being captured correctly by AutoGraph?

Some of the tests here are more integration tests (checking that statements return the correct result) rather than testing the autograph conversion explicitly.

File: frontend/test/pytest/test_autograph.py [hunk 2]
```diff
@@ -1256,6 +1256,87 @@ def f1():
...
 
+@pytest.mark.tf
+class TestLogicalOps:
+    def test_logical_basics(self, monkeypatch):
+        """Test basic logical and behavior."""
+
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        @qjit(autograph=True)
+        def f1(param):
```
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:1267]: @grwlf are there other cases we could test? E.g.,

- cases where the boolean statement is a mix of dynamic and static variables, 
- test cases where the parameters are of different types (jax arrays, booleans, numpy arrays)
- cases where the boolean is used as conditionals for if statements
- cases where the boolean is used as the conditional for while statements
- @dime10 anything else come to mind?

Comment (dime10) [reply frontend/test/pytest/test_autograph.py:1267]: That should about cover it! Maybe:
- test cases where the parameters are arbitrary Python objects which can decay to Boolean (like strings, lists, etc, anything that Python allows in an if statement)

File: frontend/test/pytest/test_autograph.py (hunk 1 above)
Comment (grwlf) [reply frontend/test/pytest/test_autograph.py:1338]: Added https://github.com/PennyLaneAI/catalyst/blob/4fd287bd97bd8328463e20d89796170729d6bda8/frontend/test/lit/test_autograph.py#L376

File: frontend/test/pytest/test_autograph.py (hunk 2 above)
Comment (grwlf) [reply frontend/test/pytest/test_autograph.py:1267]: Added. Note that in Python we can use almost any object in boolean expressions, I checked a few standard ones.

File: frontend/catalyst/ag_primitives.py [hunk 3]
```diff
@@ -411,6 +419,73 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
...
+                f"\n"
+                f"The error ocurred within the body of the following expression:\n"
+                f"{expr_info}"
+                f"\n"
+                f"To understand different types of JAX tracing errors, please refer to the "
+                f"guide at: https://jax.readthedocs.io/en/latest/errors.html\n"
+                f"\n"
+                f"If you did not intend for the conversion to happen, you may safely ignore "
+                f"this warning."
+            )
```
Comment (dime10) [inline frontend/catalyst/ag_primitives.py:466]: I'm not sure we need any warnings here. Warnings should only exist if there is something actionable you can do about them. 

With the for loop, the warning is supposed to help users change their lists into arrays for example. 
Can you think of any cases here where the user can change their code such that it would be captured by autograph?

In fact, I think the setup of this function should be slightly different than that of the for loop. With the for loop, we _want_ to convert loops since they will be more efficient afterwards. Here, we only care that the functionality _works_.
If it already works in Python (non-array inputs), great. This is probably the preferred path. But if the inputs are abstract arrays, we _need_ the jax functions.

File: frontend/test/pytest/test_autograph.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
 import pennylane as qml
 import pytest
 from numpy.testing import assert_allclose
+from pytest import MonkeyPatch
```
Comment (dime10) [inline frontend/test/pytest/test_autograph.py:27]: This import isn't needed if you aren't using the class directly, the fixture is available by default.

File: frontend/catalyst/ag_primitives.py [hunk 5]
```diff
@@ -26,6 +27,9 @@
 # as well as various utility objects.
 import pennylane as qml
 import tensorflow.python.autograph.impl.api as tf_autograph_api
+from jax.numpy import logical_and as jax_logical_and
+from jax.numpy import logical_not as jax_logical_not
+from jax.numpy import logical_or as jax_logical_or
```
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:32]: The idea is to bind functions early to detect the numpy api change which I think is both safer and easier to read in code (the one-import-per-line style is due to a autoformatter). Why do you think the module-dot-name style is better?

File: frontend/catalyst/ag_primitives.py [hunk 6]
```diff
@@ -411,6 +419,73 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
...
+    if any(map(_non_scalar_tracer, values)):
+        raise AutoGraphError(
+            "Using logical operations (and/or/not) with non-scalar tracers is not supported\n"
+            "Consider using `jax.numpy.logical_and/..logical_or/..logical_not` functions.\n"
+            f"arguments provided were: {args}"
+        )
+
+    try:
+        if _emulate_fallback_errors:
+            raise AutoGraphError("Emulated autograph fallback error")
```
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:439]: To keep the tests more predictable I think. For values like `jnp.array(True)` both paths can actually return meaningful output, we need to switch between them somehow.

File: frontend/catalyst/ag_primitives.py (hunk 3 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:466]: Can you formulate a clear criteria when to call which path? I tried to avoid making such decisions because I don't control the implementation of `jax.numpy.logical_and/or` nor do I control `__bool__` methods of random Python objects. Simply trying to call `jax.logical_and` and only call Python's `and` on failure seems to me a more responsible decision in these circumstances.

File: frontend/catalyst/ag_primitives.py [hunk 7]
```diff
@@ -411,6 +419,73 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
...
+
+    def _non_scalar_tracer(x) -> bool:
+        return isinstance(x, DynamicJaxprTracer) and reduce(lambda a, b: a * b, x.shape, 1) != 1
+
+    if any(map(_non_scalar_tracer, values)):
+        raise AutoGraphError(
+            "Using logical operations (and/or/not) with non-scalar tracers is not supported\n"
+            "Consider using `jax.numpy.logical_and/..logical_or/..logical_not` functions.\n"
+            f"arguments provided were: {args}"
+        )
```
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:435]: +1 I would also like to know this. By allowing the `jnp.array(True) and jnp.array([False])` syntax we already go beyond what the original Python allows. So how far should we go here?

File: frontend/test/pytest/test_autograph.py (hunk 4 above)
Comment (grwlf) [reply frontend/test/pytest/test_autograph.py:27]: It must be a trace of some experiments I made, thanks.

File: frontend/catalyst/ag_primitives.py (hunk 3 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:466]: Another option we might follow is to _always_ use `np.logical_and/or/not`, trying to cast all arguments to bools if they are not scalar np.arrays. @dime10 what do you think?

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:32]: Just because that's how we access all functions from the jax numpy api, even across the different PennyLane-ecosystem packages. I'd rather not have another convention that starts with `jax_*`.

File: frontend/catalyst/ag_primitives.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:439]: I don't think that's the way to do it though. If the inputs are not sufficient to control the execution path, then maybe the code logic is not designed in a sensible way?

File: frontend/catalyst/ag_primitives.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:435]: >By allowing the jnp.array(True) and jnp.array([False]) syntax we already go beyond what the original Python allows.

What do you mean? I think that example is supported with Python `and`

[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/ag_primitives.py:435]: Hey all, I just spoke to @glassnotes and the preference here is to preserve the same behaviour as JAX and NumPy -- that is, the following error should be raised with AutoGraph, and the user would need to use `jnp.all` explicitly:

```python
>>> jnp.array([0, 1]) and jnp.array([1, 2])
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
>>> jnp.all(jnp.array([0, 1]), jnp.array([1, 2]))
False
```

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:435]: This is actually more complicated than it appears. @grwlf we're still trying to understand what the best approach might be.

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:32]: I would prefer to keep `from module import function` style to bind names early. The `jax_`prefixes are optional, will remove them. Btw I wrote all this in a single line originally, it must be `isort` who made them one import per line for us.

File: frontend/catalyst/ag_primitives.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py]: No longer relevant here

File: frontend/test/pytest/test_autograph.py (hunk 4 above)
Comment (grwlf) [reply frontend/test/pytest/test_autograph.py]: Done

File: doc/changelog.md [hunk 8]
```diff
@@ -11,6 +11,10 @@
 * The AutoGraph feature, still experimental, now supports native Python `while` loops as well.
   [(#318)](https://github.com/PennyLaneAI/catalyst/pull/318)
 
+* The AutoGraph now also supports native Python `and`, `or` and `not` operators in boolean
+  expressions.
```
[the remark this answers is not in the document]
Comment (grwlf) [reply doc/changelog.md:26]: Done

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:32]: > The idea is to bind functions early to detect the numpy api change which I think is both safer and easier to read

I don't see how this detects changes in the api early, could you elaborate?

File: frontend/catalyst/ag_primitives.py (hunk 3 above)
Comment (dime10) [reply frontend/catalyst/ag_primitives.py]: For reference, we decided to only convert when both operands are abstract arrays. This is the safest option because Python Boolean operators would always fail in this instance.

No fallback should be needed then.

File: frontend/catalyst/ag_primitives.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:429]: In addition to https://github.com/PennyLaneAI/catalyst/pull/325#discussion_r1373500085, there is no need to prevent the conversion from happening when the arrays are non-scalar.

> \>>> bool(jnp.array([0, 1]) and jnp.array([1, 2]))
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

We realized the error message here is not very helpful, because it comes from the `__bool__` function being called on the LHS operand, not from ANDing together two non-scalar arrays. `jnp.all(jnp.array([0, 1]) and jnp.array([1, 2]))` for example does not work!

File: frontend/catalyst/ag_primitives.py (hunk 3 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py]: Implemented

File: frontend/catalyst/ag_primitives.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:429]: My original concern was related to the degree we want to emulate the pure-Python behavior. Without the check we have

``` python
def f(a, b):
  return a and b

>>> qjit(autograph=True)(f)(jnp.array([0,1]), jnp.array([1,1]))
array([False,  True])

>>> f(jnp.array([0,1]), jnp.array([1,1]))
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

```

File: frontend/catalyst/autograph.py [hunk 9]
```diff
@@ -66,6 +71,9 @@ def transform_ast(self, node, ctx):
         # First transform the top-level function to avoid infinite recursion.
         node = functions.transform(node, ctx)
 
+        # Convert logical expressions
+        node = logical_expressions.transform(node, ctx)
```
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/autograph.py]: Hmm. Thanks. Moved it there as well

File: frontend/catalyst/ag_primitives.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py]: fyi: The warning is removed in the current version @dime10 please check

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py]: It's extremely unlikely these functions are going to be removed, they are part of the official numpy api.

If we did this for all external functions there would be a lot more import statements across the board, I think the value add here is really low (or even negative). It's also customary across all our code base to use numpy functions as `jnp.*` (and it's variations, `np.` and `pnp.`).

File: frontend/catalyst/ag_primitives.py [hunk 10]
```diff
@@ -411,6 +419,73 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
     set_state(results)
 
 
+def _logical_op(*args, jax_fn, python_fn):
+    fallback = False
+
+    values = [f() for f in args]
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:405]: Oh interesting! I think this might allow emulating the short-circuiting behaviour of the Python `and` and `or`. Thanks :)

File: frontend/catalyst/ag_primitives.py [hunk 11]
```diff
@@ -393,6 +399,37 @@ def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_nam
     set_state(results)
 
 
+def _logical_op(*args, jax_fn, python_fn):
```
Comment (dime10) [inline frontend/catalyst/ag_primitives.py:402]: Looks great, very simple implementation 💯

**Review by dime10 — APPROVED** [review]
Looks good 👍

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py]: Implemented. Although I think that this practice weakens the code-base in general.
